File: cowork-skills/template-library/scripts/create_pptx_template.py
#!/usr/bin/env python3
"""Generate a branded PPTX reference template compatible with Pandoc.

Takes Pandoc's default reference.pptx and rebrands it with custom colors,
fonts, and styles — preserving the exact slide layout structure Pandoc expects.

Usage:
    python create_pptx_template.py [output.pptx] [--palette NAME]

Palettes: midnight (default), forest, coral, ocean, charcoal
"""

import argparse
import os
import subprocess
import sys
from pptx import Presentation
from pptx.util import Inches, Pt, Emu
from pptx.dml.color import RGBColor
from pptx.enum.text import PP_ALIGN


PALETTES = {
    "midnight": {
        "primary": RGBColor(0x1E, 0x27, 0x61),
        "secondary": RGBColor(0x2B, 0x6C, 0xB0),
        "accent": RGBColor(0x38, 0xA1, 0x69),
        "bg_dark": RGBColor(0x1E, 0x27, 0x61),
        "bg_light": RGBColor(0xF7, 0xFA, 0xFC),
        "text_dark": RGBColor(0x2D, 0x37, 0x48),
        "text_light": RGBColor(0xFF, 0xFF, 0xFF),
        "muted": RGBColor(0x71, 0x80, 0x96),
        "heading_font": "Georgia",
        "body_font": "Calibri",
    },
    "forest": {
        "primary": RGBColor(0x2C, 0x5F, 0x2D),
        "secondary": RGBColor(0x97, 0xBC, 0x62),
        "accent": RGBColor(0x38, 0xA1, 0x69),
        "bg_dark": RGBColor(0x2C, 0x5F, 0x2D),
        "bg_light": RGBColor(0xF5, 0xF5, 0xF5),
        "text_dark": RGBColor(0x1A, 0x20, 0x2C),
        "text_light": RGBColor(0xFF, 0xFF, 0xFF),
        "muted": RGBColor(0x6B, 0x72, 0x80),
        "heading_font": "Georgia",
        "body_font": "Calibri",
    },
    "coral": {
        "primary": RGBColor(0x2F, 0x3C, 0x7E),
        "secondary": RGBColor(0xF9, 0x61, 0x67),
        "accent": RGBColor(0xF9, 0xE7, 0x95),
        "bg_dark": RGBColor(0x2F, 0x3C, 0x7E),
        "bg_light": RGBColor(0xFF, 0xF5, 0xF5),
        "text_dark": RGBColor(0x2D, 0x37, 0x48),
        "text_light": RGBColor(0xFF, 0xFF, 0xFF),
        "muted": RGBColor(0x71, 0x80, 0x96),
        "heading_font": "Georgia",
        "body_font": "Calibri",
    },
    "ocean": {
        "primary": RGBColor(0x06, 0x5A, 0x82),
        "secondary": RGBColor(0x1C, 0x72, 0x93),
        "accent": RGBColor(0x21, 0x29, 0x5C),
        "bg_dark": RGBColor(0x06, 0x5A, 0x82),
        "bg_light": RGBColor(0xEB, 0xF8, 0xFF),
        "text_dark": RGBColor(0x1A, 0x20, 0x2C),
        "text_light": RGBColor(0xFF, 0xFF, 0xFF),
        "muted": RGBColor(0x64, 0x74, 0x8B),
        "heading_font": "Georgia",
        "body_font": "Calibri",
    },
    "charcoal": {
        "primary": RGBColor(0x36, 0x45, 0x4F),
        "secondary": RGBColor(0x4A, 0x55, 0x68),
        "accent": RGBColor(0x21, 0x21, 0x21),
        "bg_dark": RGBColor(0x36, 0x45, 0x4F),
        "bg_light": RGBColor(0xF2, 0xF2, 0xF2),
        "text_dark": RGBColor(0x21, 0x21, 0x21),
        "text_light": RGBColor(0xFF, 0xFF, 0xFF),
        "muted": RGBColor(0x6B, 0x72, 0x80),
        "heading_font": "Arial",
        "body_font": "Arial",
    },
}


def apply_font_to_placeholder(placeholder, font_name, font_size, font_color, bold=False):
    """Apply font styling to all runs in a placeholder's text frame."""
    if not placeholder.has_text_frame:
        return
    for paragraph in placeholder.text_frame.paragraphs:
        for run in paragraph.runs:
            run.font.name = font_name
            run.font.size = Pt(font_size)
            run.font.color.rgb = font_color
            run.font.bold = bold


def create_template(output_path, palette_name):
    p = PALETTES.get(palette_name, PALETTES["midnight"])

    # Generate Pandoc's default reference template
    base_path = "/tmp/_pandoc_base_ref.pptx"
    result = subprocess.run(
        ["pandoc", "-o", base_path, "--print-default-data-file", "reference.pptx"],
        capture_output=True
    )
    if result.returncode != 0:
        # Fallback: generate one from scratch
        subprocess.run(
            ["pandoc", "-t", "pptx", "-o", base_path],
            input=b"# Title\n\n## Slide",
            capture_output=True
        )

    prs = Presentation(base_path)

    # Modify slide masters and layouts
    for slide_master in prs.slide_masters:
        # Try to style placeholders in slide layouts
        for layout in slide_master.slide_layouts:
            for placeholder in layout.placeholders:
                ph_type = placeholder.placeholder_format.type
                if placeholder.has_text_frame:
                    for paragraph in placeholder.text_frame.paragraphs:
                        for run in paragraph.runs:
                            # Title placeholders
                            if ph_type in [1, 15]:  # TITLE, CENTERED_TITLE
                                run.font.name = p["heading_font"]
                                run.font.color.rgb = p["primary"]
                                run.font.bold = True
                            # Body placeholders
                            elif ph_type in [2, 7]:  # BODY, OBJECT
                                run.font.name = p["body_font"]
                                run.font.color.rgb = p["text_dark"]

    prs.save(output_path)
    file_size = os.path.getsize(output_path)
    print(f"Template created: {output_path} ({file_size:,} bytes, palette: {palette_name})")

    # Clean up
    if os.path.exists(base_path):
        os.remove(base_path)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Generate branded PPTX reference template")
    parser.add_argument("output", nargs="?", default="wayne_brand_template.pptx", help="Output file")
    parser.add_argument("--palette", default="midnight", choices=list(PALETTES.keys()))
    args = parser.parse_args()
    create_template(args.output, args.palette)

File: cowork-skills/template-library/scripts/create_pptx_template.js
#!/usr/bin/env node
/**
 * Generate a branded PPTX reference template.
 *
 * This creates a .pptx file with slide masters that carry Wayne's brand
 * colors, fonts, and layouts. Use it with Pandoc's --reference flag or
 * PptxGenJS's masterName to get consistent branded output every time.
 *
 * Usage:
 *   node create_pptx_template.js [output.pptx] [--palette NAME]
 *
 * Palettes: midnight (default), forest, coral, ocean, charcoal
 */

const pptxgen = require("pptxgenjs");
const path = require("path");

const PALETTES = {
  midnight: {
    primary: "1E2761", secondary: "2B6CB0", accent: "38A169",
    bg_dark: "1E2761", bg_light: "F7FAFC", text_light: "FFFFFF",
    text_dark: "2D3748", muted: "718096", border: "E2E8F0",
    headerFont: "Georgia", bodyFont: "Calibri",
  },
  forest: {
    primary: "2C5F2D", secondary: "97BC62", accent: "F5F5F5",
    bg_dark: "2C5F2D", bg_light: "F5F5F5", text_light: "FFFFFF",
    text_dark: "1A202C", muted: "6B7280", border: "D1D5DB",
    headerFont: "Georgia", bodyFont: "Calibri",
  },
  coral: {
    primary: "2F3C7E", secondary: "F96167", accent: "F9E795",
    bg_dark: "2F3C7E", bg_light: "FFF5F5", text_light: "FFFFFF",
    text_dark: "2D3748", muted: "718096", border: "E2E8F0",
    headerFont: "Georgia", bodyFont: "Calibri",
  },
  ocean: {
    primary: "065A82", secondary: "1C7293", accent: "21295C",
    bg_dark: "065A82", bg_light: "EBF8FF", text_light: "FFFFFF",
    text_dark: "1A202C", muted: "64748B", border: "CBD5E0",
    headerFont: "Georgia", bodyFont: "Calibri",
  },
  charcoal: {
    primary: "36454F", secondary: "F2F2F2", accent: "212121",
    bg_dark: "36454F", bg_light: "F2F2F2", text_light: "FFFFFF",
    text_dark: "212121", muted: "6B7280", border: "D1D5DB",
    headerFont: "Arial Black", bodyFont: "Arial",
  },
};

function createTemplate(outputPath, paletteName) {
  const p = PALETTES[paletteName] || PALETTES.midnight;
  const pres = new pptxgen();

  pres.layout = "LAYOUT_16x9";
  pres.author = "Wayne Pearce";
  pres.title = "Branded Template";

  // ── MASTER: Title Slide ──────────────────────────────
  pres.defineSlideMaster({
    title: "TITLE_SLIDE",
    background: { color: p.bg_dark },
    objects: [
      // Bottom accent bar
      { rect: { x: 0, y: 4.8, w: 10, h: 0.825, fill: { color: p.secondary } } },
      // Title placeholder
      {
        placeholder: {
          options: {
            name: "title", type: "title",
            x: 0.8, y: 1.2, w: 8.4, h: 1.8,
            fontFace: p.headerFont, fontSize: 40, color: p.text_light,
            bold: true, align: "left", valign: "middle",
          },
        },
      },
      // Subtitle placeholder
      {
        placeholder: {
          options: {
            name: "subtitle", type: "body",
            x: 0.8, y: 3.1, w: 8.4, h: 0.8,
            fontFace: p.bodyFont, fontSize: 18, color: p.text_light,
            align: "left", valign: "top",
          },
        },
      },
    ],
  });

  // ── MASTER: Content Slide ────────────────────────────
  pres.defineSlideMaster({
    title: "CONTENT_SLIDE",
    background: { color: p.bg_light },
    objects: [
      // Top accent bar
      { rect: { x: 0, y: 0, w: 10, h: 0.06, fill: { color: p.secondary } } },
      // Slide title placeholder
      {
        placeholder: {
          options: {
            name: "title", type: "title",
            x: 0.6, y: 0.25, w: 8.8, h: 0.65,
            fontFace: p.headerFont, fontSize: 26, color: p.primary,
            bold: true, align: "left", valign: "middle",
          },
        },
      },
      // Body placeholder
      {
        placeholder: {
          options: {
            name: "body", type: "body",
            x: 0.6, y: 1.1, w: 8.8, h: 4.0,
            fontFace: p.bodyFont, fontSize: 16, color: p.text_dark,
            align: "left", valign: "top",
          },
        },
      },
      // Footer
      {
        text: {
          text: "Wayne Pearce",
          options: {
            x: 0.6, y: 5.25, w: 4, h: 0.3,
            fontFace: p.bodyFont, fontSize: 8, color: p.muted,
            italic: true,
          },
        },
      },
    ],
  });

  // ── MASTER: Section Break ────────────────────────────
  pres.defineSlideMaster({
    title: "SECTION_BREAK",
    background: { color: p.primary },
    objects: [
      { rect: { x: 0.6, y: 2.9, w: 2, h: 0.05, fill: { color: p.secondary } } },
      {
        placeholder: {
          options: {
            name: "title", type: "title",
            x: 0.6, y: 1.5, w: 8.8, h: 1.3,
            fontFace: p.headerFont, fontSize: 36, color: p.text_light,
            bold: true, align: "left", valign: "bottom",
          },
        },
      },
      {
        placeholder: {
          options: {
            name: "body", type: "body",
            x: 0.6, y: 3.2, w: 8.8, h: 1.2,
            fontFace: p.bodyFont, fontSize: 16, color: p.text_light,
            align: "left", valign: "top",
          },
        },
      },
    ],
  });

  // ── MASTER: Two-Column ───────────────────────────────
  pres.defineSlideMaster({
    title: "TWO_COLUMN",
    background: { color: p.bg_light },
    objects: [
      { rect: { x: 0, y: 0, w: 10, h: 0.06, fill: { color: p.secondary } } },
      {
        placeholder: {
          options: {
            name: "title", type: "title",
            x: 0.6, y: 0.25, w: 8.8, h: 0.65,
            fontFace: p.headerFont, fontSize: 26, color: p.primary,
            bold: true, align: "left",
          },
        },
      },
      // Left column
      {
        placeholder: {
          options: {
            name: "left", type: "body",
            x: 0.6, y: 1.1, w: 4.1, h: 4.0,
            fontFace: p.bodyFont, fontSize: 14, color: p.text_dark,
            align: "left", valign: "top",
          },
        },
      },
      // Right column
      {
        placeholder: {
          options: {
            name: "right", type: "body",
            x: 5.3, y: 1.1, w: 4.1, h: 4.0,
            fontFace: p.bodyFont, fontSize: 14, color: p.text_dark,
            align: "left", valign: "top",
          },
        },
      },
    ],
  });

  // ── MASTER: Closing Slide ────────────────────────────
  pres.defineSlideMaster({
    title: "CLOSING_SLIDE",
    background: { color: p.bg_dark },
    objects: [
      { rect: { x: 0, y: 4.8, w: 10, h: 0.825, fill: { color: p.secondary } } },
      {
        placeholder: {
          options: {
            name: "title", type: "title",
            x: 0.8, y: 1.5, w: 8.4, h: 1.5,
            fontFace: p.headerFont, fontSize: 36, color: p.text_light,
            bold: true, align: "center", valign: "middle",
          },
        },
      },
      {
        placeholder: {
          options: {
            name: "body", type: "body",
            x: 0.8, y: 3.2, w: 8.4, h: 1.0,
            fontFace: p.bodyFont, fontSize: 16, color: p.text_light,
            align: "center", valign: "top",
          },
        },
      },
    ],
  });

  // ── Generate sample slides so Pandoc can pick up the masters ──
  const titleSlide = pres.addSlide({ masterName: "TITLE_SLIDE" });
  titleSlide.addText("Presentation Title", { placeholder: "title" });
  titleSlide.addText("Subtitle goes here", { placeholder: "subtitle" });

  const contentSlide = pres.addSlide({ masterName: "CONTENT_SLIDE" });
  contentSlide.addText("Content Slide", { placeholder: "title" });
  contentSlide.addText("Body text placeholder", { placeholder: "body" });

  const sectionSlide = pres.addSlide({ masterName: "SECTION_BREAK" });
  sectionSlide.addText("Section Title", { placeholder: "title" });
  sectionSlide.addText("Section description", { placeholder: "body" });

  const twoColSlide = pres.addSlide({ masterName: "TWO_COLUMN" });
  twoColSlide.addText("Two Columns", { placeholder: "title" });

  const closingSlide = pres.addSlide({ masterName: "CLOSING_SLIDE" });
  closingSlide.addText("Thank You", { placeholder: "title" });
  closingSlide.addText("Questions?", { placeholder: "body" });

  pres.writeFile({ fileName: outputPath }).then(() => {
    console.log(`Template created: ${outputPath} (palette: ${paletteName})`);
    console.log(`Masters: TITLE_SLIDE, CONTENT_SLIDE, SECTION_BREAK, TWO_COLUMN, CLOSING_SLIDE`);
  });
}

// CLI
const args = process.argv.slice(2);
let output = "wayne_brand_template.pptx";
let palette = "midnight";

for (let i = 0; i < args.length; i++) {
  if (args[i] === "--palette" && args[i + 1]) {
    palette = args[i + 1];
    i++;
  } else if (!args[i].startsWith("--")) {
    output = args[i];
  }
}

createTemplate(output, palette);
